netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,313 @@
+      }
+
+      @Override
+      public void streamInactive(Http2Stream stream) {
+        createNextPendingStream();
+      }
+    });
+  }
+
+  /**
+   * Builder for new instances of {@link BufferingHttp2ConnectionEncoder}.
+   */
+  public static class Builder extends DefaultHttp2ConnectionEncoder.Builder {
+    @Override
+    public Builder connection(Http2Connection connection) {
+      this.connection = connection;
+      return this;
+    }
+
+    @Override
+    public Builder lifecycleManager(Http2LifecycleManager lifecycleManager) {
+      this.lifecycleManager = lifecycleManager;
+      return this;
+    }
+
+    @Override
+    public Http2LifecycleManager lifecycleManager() {
+      return lifecycleManager;
+    }
+
+    @Override
+    public Builder frameWriter(Http2FrameWriter frameWriter) {
+      this.frameWriter = frameWriter;
+      return this;
+    }
+
+    @Override
+    public Http2ConnectionEncoder build() {
+      return new BufferingHttp2ConnectionEncoder(this);
+    }
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int padding, boolean endStream, ChannelPromise promise) {
+    return writeHeaders(ctx, streamId, headers, 0, Http2CodecUtil.DEFAULT_PRIORITY_WEIGHT, false,
+                        padding, endStream, promise);
+  }
+
+  @Override
+  public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers,
+                                    int streamDependency, short weight, boolean exclusive,
+                                    int padding, boolean endOfStream, ChannelPromise promise) {
+    if (existingStream(streamId)) {
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().local().acceptingNewStreams()) {
+      assert streamId > largestCreatedStreamId;
+      largestCreatedStreamId = streamId;
+      return super.writeHeaders(ctx, streamId, headers, streamDependency, weight, exclusive,
+                                padding, endOfStream, promise);
+    }
+    if (connection().isGoAway()) {
+      promise.setFailure(
+              connectionError(PROTOCOL_ERROR, "Sending headers after connection going away."));
+      return promise;
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream == null) {
+      pendingStream = new PendingStream(ctx, streamId);
+      pendingStreams.put(streamId, pendingStream);
+    }
+    pendingStream.frames.add(new HeadersFrame(headers, streamDependency, weight, exclusive,
+                             padding, endOfStream, promise));
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
```

Comment:
Ah, right. That makes sense.  Maybe add a comment so we don't confuse ourselves later? :)